File: csharp/ButtplugCSharpFFI/ButtplugClient.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Threading.Tasks;
using System.Runtime.InteropServices;

namespace Buttplug
{
    public class ButtplugClient : IDisposable
    {
        private static readonly Dictionary<uint, WeakReference> _clientStorage = new Dictionary<uint, WeakReference>();
        private static uint _clientCounter = 1;

        private readonly ButtplugFFIMessageSorter _messageSorter;
        private readonly ButtplugFFIClientHandle _clientHandle;

        /// <summary>
        /// Stores information about devices currently connected to the server.
        /// </summary>
        private readonly ConcurrentDictionary<uint, ButtplugClientDevice> _devices;
        private readonly ButtplugCallback _sorterCallbackDelegate;

        // To detect redundant calls
        private bool _disposed;
        private GCHandle _indexHandle;

        public ButtplugClientDevice[] Devices => _devices.Values.ToArray();

        /// <summary>
        /// Name of the client, used for server UI/permissions.
        /// </summary>
        public string Name { get; }
        public bool Connected { get; private set; }

        /// <summary>
        /// Event fired on Buttplug device added, either after connect or while scanning for devices.
        /// </summary>
        public event EventHandler<DeviceAddedEventArgs> DeviceAdded;

        /// <summary>
        /// Event fired on Buttplug device removed. Can fire at any time after device connection.
        /// </summary>
        public event EventHandler<DeviceRemovedEventArgs> DeviceRemoved;

        /// <summary>
        /// Fires when an error that was not provoked by a client action is received from the server,
        /// such as a device exception, message parsing error, etc... Server may possibly disconnect
        /// after this event fires.
        /// </summary>
        public event EventHandler<ButtplugExceptionEventArgs> ErrorReceived;

        /// <summary>
        /// Event fired when the server has finished scanning for devices.
        /// </summary>
        public event EventHandler ScanningFinished;

        /// <summary>
        /// Event fired when a server ping timeout has occured.
        /// </summary>
        public event EventHandler PingTimeout;

        /// <summary>
        /// Event fired when a server disconnect has occured.
        /// </summary>
        public event EventHandler ServerDisconnect;

        public bool IsScanning { get; private set; }

        private object _disposeLock;

        public ButtplugClient(string aClientName): this(aClientName, StaticSorterCallback)
        {
        }

        protected ButtplugClient(string aClientName, ButtplugCallback aCallback)
        {
            Name = aClientName;
            _sorterCallbackDelegate = aCallback;
            _disposeLock = new object();

            _messageSorter = new ButtplugFFIMessageSorter();
            _devices = new ConcurrentDictionary<uint, ButtplugClientDevice>();

            var context = new WeakReference(this);
            var clientIndex = _clientCounter;
            _clientCounter += 1;

            // Since we can pass the handle, I don't *think* this needs to be pinned?
            _indexHandle = GCHandle.Alloc(clientIndex);
            _clientStorage.Add(clientIndex, context);
            _clientHandle = ButtplugFFI.SendCreateClient(aClientName, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle));
        }

        ~ButtplugClient() => Dispose(false);

        public async Task ConnectAsync(ButtplugEmbeddedConnectorOptions aConnector)
        {
            if (aConnector == null)
            {
                aConnector = new ButtplugEmbeddedConnectorOptions();
            }

            await ButtplugFFI.SendConnectLocal(
                _messageSorter,
                _clientHandle,
                aConnector.ServerName,
                aConnector.MaxPingTime,
                aConnector.AllowRawMessages,
                aConnector.DeviceConfigJSON,
                aConnector.UserDeviceConfigJSON,
                aConnector.DeviceCommunicationManagerTypes,
                _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                .ConfigureAwait(false);

            Connected = true;
        }

        public async Task ConnectAsync(ButtplugWebsocketConnectorOptions aConnector)
        {
            await ButtplugFFI.SendConnectWebsocket(_messageSorter, _clientHandle, aConnector.NetworkAddress.ToString(), false, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                             .ConfigureAwait(false);

            Connected = true;
        }

        public async Task DisconnectAsync()
        {
            await ButtplugFFI.SendDisconnect(_messageSorter, _clientHandle, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                             .ConfigureAwait(false);

            _devices.Clear();
            Connected = false;
        }

        protected static void StaticSorterCallback(IntPtr ctx, IntPtr buf, int buf_length)
        {
            GCHandle indexHandle = GCHandle.FromIntPtr(ctx);
            uint index = (uint)indexHandle.Target;

            var client = _clientStorage[index];
            if (client.IsAlive)
            {
                ((ButtplugClient)client.Target).SorterCallbackHandler(buf, buf_length);
            }
        }

        protected void SorterCallbackHandler(IntPtr buf, int buf_length)
        {
            // Process the data BEFORE we throw to the C# executor, otherwise
            // Rust will clean up the memory and we'll have nothing to read
            // from, meaning a null message at best and a crash at worst.
            Span<byte> byteArray;
            unsafe
            {
                byteArray = new Span<byte>(buf.ToPointer(), buf_length);
            }

            var server_message = ButtplugFFIServerMessage.Parser.ParseFrom(byteArray.ToArray());
            // Run the response in the context of the C# executor, not the Rust
            // thread. This means that if something goes wrong we at least
            // aren't blocking a rust executor thread.
            Task.Run(() =>
            {
                if (server_message.Id > 0)
                {
                    _messageSorter.CheckMessage(server_message);
                }
                else if (server_message.Message.MsgCase == ButtplugFFIServerMessage.Types.FFIMessage.MsgOneofCase.ServerMessage)
                {
                    if (server_message.Message.ServerMessage.MsgCase == ServerMessage.MsgOneofCase.DeviceAdded)
                    {
                        var device_added_message = server_message.Message.ServerMessage.DeviceAdded;
                        if (_devices.ContainsKey(device_added_message.Index))
                        {
                            ErrorReceived?.Invoke(this, new ButtplugExceptionEventArgs(new ButtplugDeviceException("A duplicate device index was received. This is most likely a bug, please file at https://github.com/buttplugio/buttplug-rs-ffi")));
                            return;
                        }

                        var device_handle = ButtplugFFI.SendCreateDevice(_clientHandle, device_added_message.Index);
                        var attribute_dict = new Dictionary<ServerMessage.Types.MessageAttributeType, ButtplugMessageAttributes>();
                        for (var i = 0; i < device_added_message.MessageAttributes.Count; ++i)
                        {
                            var attributes = device_added_message.MessageAttributes[i];
                            var device_message_attributes = new ButtplugMessageAttributes(attributes.FeatureCount, attributes.StepCount.ToArray(),
                                                                                          attributes.Endpoints.ToArray(), attributes.MaxDuration.ToArray(), null, null);
                            attribute_dict.Add(attributes.MessageType, device_message_attributes);
                        }

                        var device = new ButtplugClientDevice(_messageSorter, device_handle, device_added_message.Index, device_added_message.Name, attribute_dict, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle));
                        _devices.TryAdd(device_added_message.Index, device);
                        DeviceAdded?.Invoke(this, new DeviceAddedEventArgs(device));
                    }
                    else if (server_message.Message.ServerMessage.MsgCase == ServerMessage.MsgOneofCase.DeviceRemoved)
                    {
                        var device_removed_message = server_message.Message.ServerMessage.DeviceRemoved;
                        if (!_devices.ContainsKey(device_removed_message.Index))
                        {
                            // Device was removed from our dict before we could remove it ourselves.
                            return;
                        }

                        try
                        {
                            var device = _devices[device_removed_message.Index];
                            _devices.TryRemove(device_removed_message.Index, out _);
                            DeviceRemoved?.Invoke(this, new DeviceRemovedEventArgs(device));
                        }
                        catch (KeyNotFoundException)
                        {
                            ErrorReceived?.Invoke(this, new ButtplugExceptionEventArgs(new ButtplugDeviceException($"Cannot remove device index {device_removed_message.Index}, device not found.")));
                        }
                    }
                    else if (server_message.Message.ServerMessage.MsgCase == ServerMessage.MsgOneofCase.Disconnect)
                    {
                        Connected = false;
                        _devices.Clear();
                        ServerDisconnect?.Invoke(this, null);
                    }
                    else if (server_message.Message.ServerMessage.MsgCase == ServerMessage.MsgOneofCase.Error)
                    {
                        var errorMsg = server_message.Message.ServerMessage.Error;
                        var error = ButtplugException.FromError(errorMsg);
                        if (error is ButtplugPingException)
                        {
                            PingTimeout?.Invoke(this, null);
                        }

                        ErrorReceived?.Invoke(this, new ButtplugExceptionEventArgs(error));
                    }
                    else if (server_message.Message.ServerMessage.MsgCase == ServerMessage.MsgOneofCase.ScanningFinished)
                    {
                        IsScanning = false;
                        ScanningFinished?.Invoke(this, null);
                    }
                    else
                    {
                        // We should probably do something here with unhandled events, but I'm not particularly sure what. I miss pattern matching. :(
                    }
                }
                else
                {
                    // We should probably do something here with unhandled events, but I'm not particularly sure what. I miss pattern matching. :(
                }
            });
        }

        public async Task StartScanningAsync()
        {
            IsScanning = true;
            await ButtplugFFI.SendStartScanning(_messageSorter, _clientHandle, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                             .ConfigureAwait(false);
        }

        public async Task StopScanningAsync()
        {
            IsScanning = false;
            await ButtplugFFI.SendStopScanning(_messageSorter, _clientHandle, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                             .ConfigureAwait(false);
        }

        public async Task StopAllDevicesAsync()
        {
            await ButtplugFFI.SendStopAllDevices(_messageSorter, _clientHandle, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                             .ConfigureAwait(false);
        }
        public async Task PingAsync()
        {
            await ButtplugFFI.SendPing(_messageSorter, _clientHandle, _sorterCallbackDelegate, GCHandle.ToIntPtr(_indexHandle))
                             .ConfigureAwait(false);
        }

        public void Dispose()
        {
            Dispose(true);
            // Suppress finalization.
            GC.SuppressFinalize(this);
        }

        // Protected implementation of Dispose pattern.
        protected virtual void Dispose(bool disposing)
        {
            lock (_disposeLock) 
            {
                if (_disposed)
                {
                    return;
                }

                if (disposing)
                {
                    // Dispose managed state (managed objects).
                    _clientStorage.Remove((uint)_indexHandle.Target);
                    _indexHandle.Free();
                    _clientHandle?.Dispose();
                }

                _disposed = true;
            }
        }
    }
}

File: java/lib/src/main/java/io/buttplug/ButtplugClient.java
package io.buttplug;

import com.google.protobuf.InvalidProtocolBufferException;
import com.sun.jna.Pointer;
import io.buttplug.exceptions.ButtplugDeviceException;
import io.buttplug.exceptions.ButtplugException;
import io.buttplug.exceptions.ButtplugPingException;
import io.buttplug.protos.ButtplugRsFfi.*;

import java.nio.ByteBuffer;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

/**
 * Object representing an FFI client instance.
 *
 * Make sure to close this to prevent a leak.
 */
public class ButtplugClient implements AutoCloseable {
    private static final ObjectReferenceManager<ButtplugClient> clientReferenceManager = new ObjectReferenceManager<>();
    private static final ButtplugFFI.FFICallback systemCallback = ButtplugClient::staticSystemMessageHandler;
    private final Pointer systemCallbackCtx;

    // TODO: some sort of weak reference?
    private static final ObjectReferenceManager<CompletableFuture<ButtplugFFIServerMessage.FFIMessage>> resultReferenceManager = new ObjectReferenceManager<>();
    private static final ButtplugFFI.FFICallback resultCallback = ButtplugClient::staticResultHandler;

    private Pointer pointer;

    private boolean connected = false;
    private boolean scanning = false;

    private final HashMap<Integer, ButtplugDevice> devices = new HashMap<>();

    /**
     * Name of the client, used for server UI/permissions.
     */
    public final String name;

    /**
     * Callback fired on Buttplug device added, either after connect or while scanning for devices.
     */
    public Consumer<ButtplugDevice> onDeviceAdded;

    /**
     * Callback fired on Buttplug device removed. Can fire at any time after device connection.
     */
    public Consumer<ButtplugDevice> onDeviceRemoved;

    /**
     * Fires when an error that was not provoked by a client action is received from the server,
     * such as a device exception, message parsing error, etc... Server may possibly disconnect
     * after this event fires.
     */
    public Consumer<ButtplugException> onErrorReceived;

    // Note: these uses of Runnable has nothing to do with threading.
    /**
     * Callback fired when the server has finished scanning for devices.
     */
    public Runnable onScanningFinished;

    /**
     * Callback fired when a server ping timeout has occurred.
     */
    public Runnable onPingTimeout;

    /**
     * Callback fired when a server disconnect has occurred.
     */
    public Runnable onServerDisconnect;

    public ButtplugClient(String client_name) {
        systemCallbackCtx = clientReferenceManager.add(this);
        this.pointer = ButtplugFFI.buttplug_create_protobuf_client(client_name, systemCallback, systemCallbackCtx);
        this.name = client_name;
    }

    /**
     * Frees the FFI Client. If this is not called, then it will leak.
     */
    @Override
    public void close() {
        if (pointer != null) {
            synchronized (this) {
                if (pointer != null) {
                    ButtplugFFI.buttplug_free_client(pointer);
                    this.pointer = null;
                    clientReferenceManager.remove(systemCallbackCtx);
                }
            }
        }
    }

    public boolean isConnected() {
        return connected;
    }

    public boolean isScanning() {
        return scanning;
    }

    /**
     * @return information about devices currently connected to the server.
     */
    public Map<Integer, ButtplugDevice> getDevices() {
        return Collections.unmodifiableMap(devices);
    }

    private CompletableFuture<ButtplugFFIServerMessage.FFIMessage> sendProtobufMessage(ClientMessage.FFIMessage message) {
        if (pointer == null) {
            throw new IllegalStateException("Attempt to send message when client has already been closed!");
        }

        CompletableFuture<ButtplugFFIServerMessage.FFIMessage> future = new CompletableFuture<>();

        byte[] buf = ClientMessage.newBuilder()
                .setId(0xDEAFBEEF)
                .setMessage(message)
                .build()
                .toByteArray();

        ButtplugFFI.buttplug_client_protobuf_message(pointer, buf, new ButtplugFFI.int32_t(buf.length), resultCallback, resultReferenceManager.add(future));

        return future;
    }

    private ButtplugDevice createDevice(ServerMessage.DeviceAdded msg) {
        if (pointer == null) {
            throw new IllegalStateException("Attempt to create device when client has already been closed!");
        }
        return new ButtplugDevice(pointer, msg);
    }

    public CompletableFuture<Void> connect(EmbeddedConnectorOptions options) {
        int manager_types = 0;
        for (EmbeddedConnectorOptions.DeviceCommunicationManager val: options.deviceCommunicationManagerTypes) {
            manager_types |= val.value.getNumber();
        }

        ClientMessage.FFIMessage.Builder builder = ClientMessage.FFIMessage.newBuilder();
        builder.getConnectLocalBuilder()
                .setServerName(options.serverName)
                .setMaxPingTime(options.maxPingTime)
                .setAllowRawMessages(options.allowRawMessages)
                .setDeviceConfigurationJson(options.deviceConfigJSON)
                .setUserDeviceConfigurationJson(options.userDeviceConfigJSON)
                .setCommManagerTypes(manager_types);

        return sendProtobufMessage(builder.build())
                .thenAccept(ButtplugProtoUtil::to_result)
                .thenRun(() -> connected = true);
    }

    public CompletableFuture<Void> connect(WebsocketConnectorOptions options) {
        ClientMessage.FFIMessage.Builder builder = ClientMessage.FFIMessage.newBuilder();
        builder.getConnectWebsocketBuilder()
                .setAddress(options.networkAddress.toString())
                // TODO: allow this to be configured?
                .setBypassCertVerification(false);

        return sendProtobufMessage(builder.build())
                .thenAccept(ButtplugProtoUtil::to_result)
                .thenRun(() -> connected = true);
    }

    public CompletableFuture<Void> disconnect() {
        ClientMessage.FFIMessage message = ClientMessage.FFIMessage.newBuilder()
                .setDisconnect(ClientMessage.Disconnect.getDefaultInstance())
                .build();

        return sendProtobufMessage(message)
                .thenAccept(ButtplugProtoUtil::to_result)
                .thenRun(() -> {
                    connected = false;
                    // devices.clear()
                });
    }

    public CompletableFuture<Void> startScanning() {
        scanning = true;
        ClientMessage.FFIMessage message = ClientMessage.FFIMessage.newBuilder()
                .setStartScanning(ClientMessage.StartScanning.getDefaultInstance())
                .build();

        return sendProtobufMessage(message)
                .thenAccept(ButtplugProtoUtil::to_result);
    }

    public CompletableFuture<Void> stopScanning() {
        scanning = false;
        ClientMessage.FFIMessage message = ClientMessage.FFIMessage.newBuilder()
                .setStopScanning(ClientMessage.StopScanning.getDefaultInstance())
                .build();

        return sendProtobufMessage(message)
                .thenAccept(ButtplugProtoUtil::to_result);
    }

    public CompletableFuture<Void> stopAllDevices() {
        ClientMessage.FFIMessage message = ClientMessage.FFIMessage.newBuilder()
                .setStopAllDevices(ClientMessage.StopAllDevices.getDefaultInstance())
                .build();

        return sendProtobufMessage(message)
                .thenAccept(ButtplugProtoUtil::to_result);
    }

    public CompletableFuture<Void> ping() {
        ClientMessage.FFIMessage message = ClientMessage.FFIMessage.newBuilder()
                .setPing(ClientMessage.Ping.getDefaultInstance())
                .build();

        return sendProtobufMessage(message)
                .thenAccept(ButtplugProtoUtil::to_result);
    }

    private static void staticSystemMessageHandler(Pointer ctx, Pointer ptr, ButtplugFFI.uint32_t len) {
        ButtplugClient client = clientReferenceManager.get(ctx);

        if (client != null) {
            client.handleSystemMessage(ptr.getByteBuffer(0, len.intValue()));
        } else {
            // TODO: what do we do if we receive a call from rust after we have freed the client?
        }
    }

    private void handleSystemMessage(ByteBuffer buf) {
        try {
            ButtplugFFIServerMessage incoming = ButtplugFFIServerMessage.parseFrom(buf);
            // TODO: is there another way we want to do this?
            // Run the response in the context of the Java executor, not the Rust
            // thread. This means that if something goes wrong we at least
            // aren't blocking a rust executor thread.
            CompletableFuture.runAsync(() -> {
                if (incoming.getId() != 0) {
                    // TODO: somehow got non-system message, warn/error?
                    return;
                }

                ButtplugFFIServerMessage.FFIMessage incomingMessage = incoming.getMessage();

                if (incomingMessage.hasServerMessage()) {
                    ServerMessage serverMsg = incomingMessage.getServerMessage();
                    if (serverMsg.hasDeviceAdded()) {
                        ServerMessage.DeviceAdded msg = serverMsg.getDeviceAdded();

                        if (devices.containsKey(msg.getIndex())) {
                            if (onErrorReceived != null) {
                                onErrorReceived.accept(new ButtplugDeviceException("A duplicate device index was received. This is most likely a bug, please file at https://github.com/buttplugio/buttplug-rs-ffi"));
                            }
                            return;
                        }

                        ButtplugDevice device = createDevice(msg);
                        devices.put(msg.getIndex(), device);
                        if (onDeviceAdded != null) {
                            onDeviceAdded.accept(device);
                        }
                    } else if (serverMsg.hasDeviceRemoved()) {
                        ServerMessage.DeviceRemoved msg = serverMsg.getDeviceRemoved();

                        if (!devices.containsKey(msg.getIndex())) {
                            // Device was removed from our map before we could remove it ourselves.
                            return;
                        }

                        ButtplugDevice device = devices.remove(msg.getIndex());
                        if (onDeviceRemoved != null) {
                            onDeviceRemoved.accept(device);
                        }
                        device.close();
                    } else if (serverMsg.hasDisconnect()) {
                        connected = false;
                        devices.clear();
                        if (onServerDisconnect != null) {
                            onServerDisconnect.run();
                        }
                    } else if (serverMsg.hasScanningFinished()) {
                        scanning = false;
                        if (onScanningFinished != null) {
                            onScanningFinished.run();
                        }
                    } else if (serverMsg.hasError()) {
                        ButtplugException ex = ButtplugException.fromError(serverMsg.getError());
                        if (ex instanceof ButtplugPingException) {
                            if (onPingTimeout != null) {
                                onPingTimeout.run();
                            }
                            // return?
                        }

                        if (onErrorReceived != null) {
                            onErrorReceived.accept(ex);
                        }
                    } else {
                        // unhandled event?
                        return;
                    }
                } else {
                    // unhandled event?
                    return;
                }
            });
        } catch (InvalidProtocolBufferException ignored) {
            // TODO: log warning/error? Shouldn't really happen.
        }
    }

    private static void staticResultHandler(Pointer ctx, Pointer ptr, ButtplugFFI.uint32_t len) {
        CompletableFuture<ButtplugFFIServerMessage.FFIMessage> future = resultReferenceManager.get(ctx);

        try {
            ButtplugProtoUtil.protobufResultHandler(future, ptr, len.intValue());
        } finally {
            resultReferenceManager.remove(ctx);
        }
    }
}
